packages/babel-traverse/src/path/conversion.js

```diff
@@ -25,14 +25,415 @@ export function ensureBlock() {
+    return t.arrowFunctionExpression([t.restElement(argsBinding)], t.callExpression(t.super(), [
+      t.spreadElement(t.identifier(argsBinding.name)),
+    ]));
+  });
+}
+
+// Create a binding for a function that will call "super.foo()" or "super[foo]()".
+function getSuperPropCallBinding(thisEnvFn, propName) {
+  return getBinding(thisEnvFn, `superprop_call:${propName || ""}`, () => {
+    const argsBinding = thisEnvFn.scope.generateUidIdentifier("args");
+    const argsList = [ t.restElement(argsBinding) ];
+
+    let fnBody;
+    if (propName) {
+      // (...args) => super.foo(...args)
+      fnBody = t.callExpression(
+        t.memberExpression(t.super(), t.identifier(propName)),
+        [ t.spreadElement(t.identifier(argsBinding.name)) ],
+      );
+    } else {
+      const method = thisEnvFn.scope.generateUidIdentifier("prop");
+      // (method, ...args) => super[method](...args)
+      argsList.unshift(method);
+      fnBody = t.callExpression(
+        t.memberExpression(t.super(), t.identifier(method.name), true /* computed */),
+        [ t.spreadElement(t.identifier(argsBinding.name)) ],
+      );
+    }
+
+    return t.arrowFunctionExpression(argsList, fnBody);
+  });
+}
+
+// Create a binding for a function that will call "super.foo" or "super[foo]".
+function getSuperPropBinding(thisEnvFn, isAssignment, propName) {
+  const op = isAssignment ? "set" : "get";
+
+  return getBinding(thisEnvFn, `superprop_${op}:${propName || ""}`, () => {
+    const argsList = [];
+
+    let fnBody;
+    if (propName) {
+      // () => super.foo
+      fnBody = t.memberExpression(t.super(), t.identifier(propName));
+    } else {
+      const method = thisEnvFn.scope.generateUidIdentifier("prop");
+      // (method) => super[method]
+      argsList.unshift(method);
+      fnBody = t.memberExpression(t.super(), t.identifier(method.name), true /* computed */);
+    }
+
+    if (isAssignment) {
+      const valueIdent = thisEnvFn.scope.generateUidIdentifier("value");
+      argsList.push(valueIdent);
+
+      fnBody = t.assignmentExpression("=", fnBody, t.identifier(valueIdent.name));
+    }
+
+    return t.arrowFunctionExpression(argsList, fnBody);
+  });
+}
+
+function getBinding(thisEnvFn, key, init) {
+  const cacheKey = "binding:" + key;
+  let data = thisEnvFn.getData(cacheKey);
+  if (!data) {
+    const id = thisEnvFn.scope.generateUidIdentifier(key);
+    data = id.name;
+    thisEnvFn.setData(cacheKey, data);
+
+    thisEnvFn.scope.push({
+      id: id,
+      init: init(data),
+    });
+  }
+
+  return data;
+}
+
+function getScopeInformation(fnPath) {
+  const thisPaths = [];
+  const argumentsPaths = [];
+  const newTargetPaths = [];
+  const superProps = [];
+  const superCalls = [];
+
+  fnPath.traverse({
+    Function(child) {
+      if (child.isArrowFunctionExpression() || child.isClassProperty()) return;
+      child.skip();
+    },
+    ThisExpression(child) {
+      thisPaths.push(child);
+    },
+    JSXIdentifier(child) {
```

Comment:
It is, I guess I should make this `JSXMemberExpression` now that you mention it though.